File: frontend/src/types/candidate.ts
/**
 * Strongly-typed definitions for Candidate Domain Module.
 */

export type ProficiencyLevel = "BEGINNER" | "INTERMEDIATE" | "ADVANCED" | "EXPERT";

export interface CandidateProfile {
  id: string;
  user_id: string;
  full_name: string;
  email: string;
  headline: string | null;
  bio: string | null;
  current_role: string | null;
  experience_years: number;
  education_level: string | null;
  location_city: string | null;
  location_state: string | null;
  resume_filename: string | null;
  resume_file_size: number | null;
  resume_uploaded_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface CandidateProfileUpdateData {
  full_name?: string;
  headline?: string | null;
  bio?: string | null;
  current_role?: string | null;
  experience_years?: number | null;
  education_level?: string | null;
  location_city?: string | null;
  location_state?: string | null;
}

export interface CandidateSkill {
  id: string;
  candidate_id: string;
  skill_id: string;
  skill_name: string;
  category: string;
  skill_type: string;
  proficiency: ProficiencyLevel;
  years_experience: number;
  is_verified: boolean;
  created_at: string;
  updated_at: string;
}

export interface CandidateSkillCreateData {
  skill_id: string;
  proficiency: ProficiencyLevel;
  years_experience?: number;
}

export interface CandidateSkillUpdateData {
  proficiency?: ProficiencyLevel;
  years_experience?: number;
}

export interface CandidateEducation {
  id: string;
  candidate_id: string;
  institution: string;
  degree: string;
  field_of_study: string | null;
  start_year: number | null;
  end_year: number | null;
  is_current: boolean;
  grade: string | null;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface CandidateEducationCreateData {
  institution: string;
  degree: string;
  field_of_study?: string | null;
  start_year?: number | null;
  end_year?: number | null;
  is_current?: boolean;
  grade?: string | null;
  description?: string | null;
}

export interface CandidateEducationUpdateData {
  institution?: string;
  degree?: string;
  field_of_study?: string | null;
  start_year?: number | null;
  end_year?: number | null;
  is_current?: boolean;
  grade?: string | null;
  description?: string | null;
}

export interface CandidateExperience {
  id: string;
  candidate_id: string;
  company: string;
  title: string;
  employment_type: string | null;
  location: string | null;
  start_date: string | null;
  end_date: string | null;
  is_current: boolean;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface CandidateExperienceCreateData {
  company: string;
  title: string;
  employment_type?: string | null;
  location?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  is_current?: boolean;
  description?: string | null;
}

export interface CandidateExperienceUpdateData {
  company?: string;
  title?: string;
  employment_type?: string | null;
  location?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  is_current?: boolean;
  description?: string | null;
}

export interface CandidateResumeUploadData {
  filename: string;
  file_size: number;
  resume_text?: string | null;
}

export interface CandidateResumeResponse {
  filename: string | null;
  file_size: number | null;
  uploaded_at: string | null;
  has_resume: boolean;
}

export interface ProfileCompleteness {
  percentage: number;
  completed_sections: string[];
  missing_sections: string[];
  section_scores: Record<string, number>;
}

export interface CandidateDashboardData {
  profile: CandidateProfile;
  completeness: ProfileCompleteness;
  skills_count: number;
  skills_by_proficiency: Record<ProficiencyLevel, number>;
  experience_count: number;
  education_count: number;
  recent_experiences: CandidateExperience[];
  highest_education: CandidateEducation | null;
}

File: backend/app/schemas/candidate.py
"""Pydantic schemas for the candidate domain module."""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.candidate_skill import ProficiencyLevel


# ---------------------------------------------------------------------------
# Candidate Profile Schemas
# ---------------------------------------------------------------------------
class CandidateProfileRead(BaseModel):
    """Detailed candidate profile read schema."""

    id: uuid.UUID
    user_id: uuid.UUID
    full_name: str
    email: str
    headline: str | None = None
    bio: str | None = None
    current_role: str | None = None
    experience_years: float = 0.0
    education_level: str | None = None
    location_city: str | None = None
    location_state: str | None = None
    resume_filename: str | None = None
    resume_file_size: int | None = None
    resume_uploaded_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CandidateProfileUpdate(BaseModel):
    """Update schema for candidate profile."""

    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    headline: str | None = Field(default=None, max_length=255)
    bio: str | None = None
    current_role: str | None = Field(default=None, max_length=100)
    experience_years: float | None = Field(default=None, ge=0.0)
    education_level: str | None = Field(default=None, max_length=100)
    location_city: str | None = Field(default=None, max_length=100)
    location_state: str | None = Field(default=None, max_length=100)


# ---------------------------------------------------------------------------
# Candidate Skill Schemas
# ---------------------------------------------------------------------------
class CandidateSkillCreate(BaseModel):
    """Request schema for associating a canonical skill with candidate."""

    skill_id: uuid.UUID
    proficiency: ProficiencyLevel = ProficiencyLevel.INTERMEDIATE
    years_experience: float = Field(default=0.0, ge=0.0)


class CandidateSkillUpdate(BaseModel):
    """Request schema for updating proficiency or experience of an attached skill."""

    proficiency: ProficiencyLevel | None = None
    years_experience: float | None = Field(default=None, ge=0.0)


class CandidateSkillResponse(BaseModel):
    """Response schema for a candidate-linked skill."""

    id: uuid.UUID
    candidate_id: uuid.UUID
    skill_id: uuid.UUID
    skill_name: str
    category: str
    skill_type: str
    proficiency: ProficiencyLevel
    years_experience: float
    is_verified: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Candidate Education Schemas
# ---------------------------------------------------------------------------
class CandidateEducationCreate(BaseModel):
    """Request schema for adding education entry."""

    institution: str = Field(..., min_length=1, max_length=255)
    degree: str = Field(..., min_length=1, max_length=100)
    field_of_study: str | None = Field(default=None, max_length=100)
    start_year: int | None = Field(default=None, ge=1900, le=2100)
    end_year: int | None = Field(default=None, ge=1900, le=2100)
    is_current: bool = False
    grade: str | None = Field(default=None, max_length=50)
    description: str | None = None


class CandidateEducationUpdate(BaseModel):
    """Request schema for modifying an education entry."""

    institution: str | None = Field(default=None, min_length=1, max_length=255)
    degree: str | None = Field(default=None, min_length=1, max_length=100)
    field_of_study: str | None = Field(default=None, max_length=100)
    start_year: int | None = Field(default=None, ge=1900, le=2100)
    end_year: int | None = Field(default=None, ge=1900, le=2100)
    is_current: bool | None = None
    grade: str | None = Field(default=None, max_length=50)
    description: str | None = None


class CandidateEducationResponse(BaseModel):
    """Response schema for an education entry."""

    id: uuid.UUID
    candidate_id: uuid.UUID
    institution: str
    degree: str
    field_of_study: str | None = None
    start_year: int | None = None
    end_year: int | None = None
    is_current: bool
    grade: str | None = None
    description: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Candidate Experience Schemas
# ---------------------------------------------------------------------------
class CandidateExperienceCreate(BaseModel):
    """Request schema for adding professional experience."""

    company: str = Field(..., min_length=1, max_length=255)
    title: str = Field(..., min_length=1, max_length=100)
    employment_type: str | None = Field(default=None, max_length=50)
    location: str | None = Field(default=None, max_length=100)
    start_date: str | None = Field(default=None, max_length=50)
    end_date: str | None = Field(default=None, max_length=50)
    is_current: bool = False
    description: str | None = None


class CandidateExperienceUpdate(BaseModel):
    """Request schema for updating professional experience."""

    company: str | None = Field(default=None, min_length=1, max_length=255)
    title: str | None = Field(default=None, min_length=1, max_length=100)
    employment_type: str | None = Field(default=None, max_length=50)
    location: str | None = Field(default=None, max_length=100)
    start_date: str | None = Field(default=None, max_length=50)
    end_date: str | None = Field(default=None, max_length=50)
    is_current: bool | None = None
    description: str | None = None


class CandidateExperienceResponse(BaseModel):
    """Response schema for professional experience."""

    id: uuid.UUID
    candidate_id: uuid.UUID
    company: str
    title: str
    employment_type: str | None = None
    location: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    is_current: bool
    description: str | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Resume Schemas
# ---------------------------------------------------------------------------
class CandidateResumeUploadRequest(BaseModel):
    """Resume metadata and optional content upload."""

    filename: str = Field(..., min_length=1, max_length=255)
    file_size: int = Field(..., ge=0)
    resume_text: str | None = None


class CandidateResumeResponse(BaseModel):
    """Resume metadata response."""

    filename: str | None = None
    file_size: int | None = None
    uploaded_at: datetime | None = None
    has_resume: bool = False


# ---------------------------------------------------------------------------
# Profile Completeness & Dashboard Schemas
# ---------------------------------------------------------------------------
class ProfileCompletenessResponse(BaseModel):
    """Deterministic profile completeness score and section breakdown."""

    percentage: int
    completed_sections: list[str]
    missing_sections: list[str]
    section_scores: dict[str, int]


class CandidateDashboardResponse(BaseModel):
    """Aggregated metrics for candidate dashboard."""

    profile: CandidateProfileRead
    completeness: ProfileCompletenessResponse
    skills_count: int
    skills_by_proficiency: dict[str, int]
    experience_count: int
    education_count: int
    recent_experiences: list[CandidateExperienceResponse]
    highest_education: CandidateEducationResponse | None = None
